# Q&A Model based on Custom Data using GooglePalm LLM Model

In [ ]:
from langchain.llms import GooglePalm
api_key="AIzaSyDWISyWAdPJlm57KB3Vcp2nYqGA0rta-XU"
llm=GooglePalm(google_api_key=api_key,temperature=.7)

### Testing the GooglePalm LLM model

In [44]:
essay = llm("write email requesting refund for electronic item")
print(essay)

Dear [Customer Service Representative name],

I am writing to request a refund for the [product name] that I purchased on [date]. I am not satisfied with the product because [reason for dissatisfaction].

I have attached a copy of my receipt and a picture of the product. I have also included a brief description of the problem that I am having with the product.

I would appreciate it if you would process my refund as soon as possible. I would like to receive a refund in the form of a credit to my original payment method.

Thank you for your time and attention to this matter.

Sincerely,
[Your name]


### Now let's load data from Codebasics FAQ csv file

In [32]:
from langchain.document_loaders.csv_loader import CSVLoader
loader=CSVLoader(file_path='codebasics_faqs.csv',source_column='prompt')
data=loader.load()

# Hugging Face Embeddings

In [33]:
from InstructorEmbedding import INSTRUCTOR

In [34]:
from langchain.embeddings import HuggingFaceInstructEmbeddings
embeddings = HuggingFaceInstructEmbeddings(model_name= "hkunlp/instructor-large",  model_kwargs={"device": "cpu"},query_instruction="Represent the query for retrieval: ")

load INSTRUCTOR_Transformer


Some weights of the model checkpoint at C:\Users\rohan/.cache\torch\sentence_transformers\hkunlp_instructor-large\ were not used when initializing T5EncoderModel: ['vision_model.encoder.layers.1.mlp.fc1.weight', 'vision_model.encoder.layers.11.self_attn.q_proj.bias', 'vision_model.encoder.layers.9.self_attn.k_proj.bias', 'vision_model.encoder.layers.8.mlp.fc1.weight', 'text_model.encoder.layers.0.self_attn.v_proj.weight', 'vision_model.encoder.layers.18.layer_norm2.weight', 'text_model.encoder.layers.9.self_attn.out_proj.bias', 'vision_model.encoder.layers.6.self_attn.out_proj.bias', 'text_model.encoder.layers.8.mlp.fc2.bias', 'vision_model.encoder.layers.10.self_attn.q_proj.bias', 'text_model.encoder.layers.6.mlp.fc1.weight', 'text_model.encoder.layers.6.layer_norm1.weight', 'vision_model.encoder.layers.2.layer_norm2.weight', 'text_model.encoder.layers.9.layer_norm2.bias', 'vision_model.encoder.layers.5.self_attn.q_proj.bias', 'text_model.encoder.layers.4.self_attn.out_proj.weight', '

Some weights of T5EncoderModel were not initialized from the model checkpoint at C:\Users\rohan/.cache\torch\sentence_transformers\hkunlp_instructor-large\ and are newly initialized: ['encoder.block.1.layer.0.SelfAttention.k.weight', 'encoder.block.11.layer.1.layer_norm.weight', 'encoder.block.8.layer.0.SelfAttention.o.weight', 'encoder.block.19.layer.0.layer_norm.weight', 'encoder.block.0.layer.0.SelfAttention.relative_attention_bias.weight', 'encoder.block.3.layer.0.SelfAttention.o.weight', 'encoder.block.3.layer.0.SelfAttention.k.weight', 'encoder.block.8.layer.1.DenseReluDense.wi.weight', 'encoder.block.22.layer.1.DenseReluDense.wo.weight', 'encoder.block.10.layer.0.SelfAttention.o.weight', 'encoder.block.22.layer.0.SelfAttention.k.weight', 'encoder.block.6.layer.0.layer_norm.weight', 'encoder.block.23.layer.0.layer_norm.weight', 'encoder.block.11.layer.0.SelfAttention.o.weight', 'encoder.block.2.layer.0.SelfAttention.k.weight', 'encoder.block.5.layer.0.SelfAttention.q.weight', 'en

max_seq_length  512


In [35]:
e=embeddings.embed_query("What is your refund policy")

In [45]:
e[:5]

[0.04671666771173477,
 -0.004087879788130522,
 -0.015315327793359756,
 0.029575854539871216,
 0.05121871083974838]

In [46]:
len(e)

768

As you can see above, embedding for a sentance "What is your refund policy" is a list of size 768. Looking at the numbers in this list, doesn't give any intuitive understanding of what it is but just assume that these numbers are capturing the meaning of "What is your refund policy". If you are curious to know about embeddings, go to youtube and search "codebasics word embeddings" and you will find bunch of videos with simple, intuitive explanations

### Vector store using FAISS

In [36]:
from langchain.vectorstores import FAISS
!pip install faiss-cpu

In [37]:
vectordb=FAISS.from_documents(embedding=embeddings,documents=data)

In [38]:
retreiver=vectordb.as_retriever()
rdocs=retreiver.get_relevant_documents("for how long is this course valid?")
rdocs

[Document(page_content='prompt: Once purchased, is this course available for lifetime access?\nresponse: Yes', metadata={'source': 'Once purchased, is this course available for lifetime access?', 'row': 22}),
 Document(page_content='prompt: why row and value option is not showing for the visual in PowerBI , any setting need to be change, please let me know?\nresponse: You have selected Table Visual instead of Matrix. That is why you are seeing a different interface.', metadata={'source': 'why row and value option is not showing for the visual in PowerBI , any setting need to be change, please let me know?', 'row': 46}),
 Document(page_content='prompt: Can I add this course to my resume? If Yes, how?\nresponse: Absolutely, we have a section in this course explaining how you can add the learnings from this course in your resume that will appeal to the hiring team.', metadata={'source': 'Can I add this course to my resume? If Yes, how?', 'row': 34}),
 Document(page_content='prompt: I use 

### Create RetrievalQA chain along with prompt template 🚀

In [39]:
from langchain.prompts import PromptTemplate
prompt_template="""Given the following context and a question, generate an answer based on this contest only.If the answer
is not found in the context, kindly state "I don't know." Don't try to makeup an answer.

CONTEXT: {context}
QUESTION: {question}"""

PROMPT=PromptTemplate(template=prompt_template,input_variables=['context','question'])

In [40]:
from langchain.chains import RetrievalQA
chain=RetrievalQA.from_chain_type(
    chain_type='stuff',
    retriever=retreiver,
    llm=llm,
    input_key='query',
    return_source_documents=True,
    chain_type_kwargs={'prompt':PROMPT}
)

### We are all set 👍🏼 Let's ask some questions now

In [41]:
chain("Do you guys provide internships and offer EMI payments?")

{'query': 'Do you guys provide internships and offer EMI payments?',
 'result': 'Yes, we provide internships and offer EMI payments.',
 'source_documents': [Document(page_content='prompt: Do you provide any virtual internship?\nresponse: Yes', metadata={'source': 'Do you provide any virtual internship?', 'row': 14}),
  Document(page_content="prompt: whenever I am loading fact_sales_monthly table in Power query not all columns are visible\nresponse: Have you taken SQL course in Codebasics first ?\nIf yes, delete the existing databases in MySQL workbench and re-import again. After this go to 'Home' tab in Power Query and click on Refresh button and see if it works any.\n\nCheck this reference too: https://discordapp.com/channels/1090613684163850280/1111180401478746163/1111190746507251732", metadata={'source': 'whenever I am loading fact_sales_monthly table in Power query not all columns are visible', 'row': 60}),
  Document(page_content='prompt: What if I don’t like this bootcamp?\nrespo

In [43]:
chain("Do know capital of India?")

{'query': 'Do know capital of India?',
 'result': "I don't know",
 'source_documents': [Document(page_content='prompt: What is the duration of this bootcamp? How long will it last?\nresponse: You can complete all courses in 3 months if you dedicate 2-3 hours per day.', metadata={'source': 'What is the duration of this bootcamp? How long will it last?', 'row': 8}),
  Document(page_content='prompt: Why we use Net error in place of absolute net error\nresponse: Directional Insight: The net error metric offers valuable information regarding the direction of the error in forecasting, distinguishing between positive (overestimation) and negative (underestimation) values. This enables the identification of specific products that consistently outperform or underperform relative to the expected values. Additionally, net error provides insights into the risk factor, indicating if there are instances of stockouts or excessive inventory levels.', metadata={'source': 'Why we use Net error in place 